cmd/snap-update-ns/change.go

```diff
@@ -56,10 +57,116 @@ func (c Change) String() string {
 var (
 	sysMount   = syscall.Mount
 	sysUnmount = syscall.Unmount
+	osLstat    = os.Lstat
+	// NOTE: we're not using os.MkdirAll as it is not careful about symlinks.
+	osMkdirAll = SecureMkdirAll
+	osChown    = os.Chown
 )
 
 const unmountNoFollow = 8
 
+const AT_FDCWD = -100 // not available through syscall
+
+// SecureMkdirAll is the secure variant of os.MkdirAll.
+//
+// Unlike the regular version this implementation does not follow any symbolic
+// links. At all times the new directory segment is created using mkdirat(2)
+// while holding an open file descriptor to the parent directory.
+//
+// The only handled error is mkdirat(2) that fails with EEXIST. All other
+// errors are fatal but there is no attempt to undo anything that was created.
+func SecureMkdirAll(name string, perm os.FileMode) error {
```

Comment:
could this live in `osutil` instead?

Reply:
Yes but @jdstrand requested that we limit imports from other modules. I can move it there if he agrees.